chiron/multistate.py

```diff
@@ -0,0 +1,701 @@
+        # Propagate using the mcmc sampler
+        # NOTE this needs to be updated to support neighborlists
+        (
+            self._sampler_states[replica_id],
+            self._thermodynamic_states[thermodynamic_state_id],
+            nbr_list,
+        ) = mcmc_sampler.run(sampler_state, thermodynamic_state)
+        # Append the new state to the trajectory for analysis.
+        self._traj[replica_id].append(self._sampler_states[replica_id].positions)
+
+    def _perform_swap_proposals(self):
+        """
+        Perform swap proposals between replicas.
+
+        This method should be overridden by subclasses to implement specific swapping algorithms.
+
+        Returns
+        -------
+        np.ndarray
+            An array of updated thermodynamic state indices for each replica.
+        """
+        # Placeholder implementation, should be overridden by subclasses
+        # For this example, we'll just return the current state indices
+        return self._replica_thermodynamic_states
+
+    def _mix_replicas(self) -> np.ndarray:
+        """
+        Propose and execute swaps between replicas to enhance sampling efficiency.
+
+        This method handles the logic for proposing swaps between different thermodynamic states
+        of the replicas. The specifics of the swapping algorithm depend on subclass implementations.
+
+        Returns
+        -------
+        np.ndarray
+            An array of updated thermodynamic state indices for each replica after swapping.
+        """
+        from loguru import logger as log
+
+        log.debug("Mixing replicas (does nothing for MultiStateSampler)...")
+
+        # Reset swap attempt counters for this iteration.
+        self._n_accepted_matrix[:, :] = 0
+        self._n_proposed_matrix[:, :] = 0
+
+        # Perform the swap proposals and acceptances.
+        new_replica_states = self._perform_swap_proposals()
+
+        # Calculate swap acceptance statistics
+        n_swaps_proposed = self._n_proposed_matrix.sum()
+        n_swaps_accepted = self._n_accepted_matrix.sum()
+        swap_fraction_accepted = 0.0
+        swap_fraction_accepted = (
+            n_swaps_accepted / n_swaps_proposed if n_swaps_proposed > 0 else 0.0
+        )
+        log.debug(
+            f"Accepted {n_swaps_accepted}/{n_swaps_proposed} attempted swaps ({swap_fraction_accepted * 100.0:.1f}%)"
+        )
+        return new_replica_states
+
+    def _propagate_replicas(self) -> None:
+        """
+        Propagate all replicas through their respective MCMC moves.
+
+        This method applies the corresponding MCMC move to each replica based on its
+        current thermodynamic state, thus advancing the state of each replica.
+        """
+        from loguru import logger as log
+
+        log.debug("Propagating all replicas...")
+
+        # Iterate over all replicas and propagate each one.
+        for replica_id in range(self.n_replicas):
+            self._propagate_replica(replica_id)
```

Comment:
Jax pmap can be multihost aware (I think I looked older documentation!), but as we discussed today, we might not want to use this for communication; it might be much more straightforward to use mpi for communication, but we'll have to look at this.  I created a place holder issue and we can start discussing this. #36